File: more-examples-pageobjects/java-testng/src/main/java/pageObjects/BookingDetailsPage.java
package pageObjects;

import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.interactions.Actions;
import org.openqa.selenium.support.FindBy;
import org.openqa.selenium.support.PageFactory;

public class BookingDetailsPage {

    private WebDriver webDriver;

    @FindBy(name = "firstname")
    private WebElement firstName;

    @FindBy(name = "lastname")
    private WebElement lastName;

    @FindBy(name = "phone")
    private WebElement phone;

    @FindBy(name = "email")
    private WebElement email;

    @FindBy(name = "confirmemail")
    private WebElement confirmEmail;

    @FindBy(name = "address")
    private WebElement address;

    @FindBy(name = "additionalnotes")
    private WebElement additionalNotes;

    @FindBy(name = "guest")
    private WebElement confirmBooking;

    public BookingDetailsPage(WebDriver webDriver) {
        this.webDriver = webDriver;
        PageFactory.initElements(this.webDriver, this);
    }

    public void setFirstName(String firstName) {
        this.firstName.sendKeys(firstName);
    }

    public void setLastName(String lastName) {
        this.lastName.sendKeys(lastName);
    }

    public void setPhone(String phone) {
        this.phone.sendKeys(phone);
    }

    public void setEmail(String email) {
        this.email.sendKeys(email);
    }

    public void setConfirmEmail(String email) {
        this.confirmEmail.sendKeys(email);
    }

    public void setAddress(String address) {
        this.address.sendKeys(address);
    }

    public void setAdditionalNotes(String notes) {
        this.additionalNotes.sendKeys(notes);
    }

    public BookingConfirmationPage clickConfirmBooking() {
        if (!confirmBooking.isDisplayed()) {
            Actions actions = new Actions(webDriver);
            actions.moveToElement(confirmBooking);
            actions.perform();
        }
        confirmBooking.click();
        return new BookingConfirmationPage(this.webDriver);
    }




}

File: page-objects/js-mocha-chai/complete/pages/booking-details-page.js
var webDriver = require('selenium-webdriver');
var By = webDriver.By;

var BookingConfirmationPage = require('./booking-confirmation-page.js');

BookingDetailsPage = function BookingDetailsPage(driver) {
    this.driver = driver;
    this.firstName = By.name("firstname");
    this.lastName = By.name("lastname");
    this.phone = By.name("phone");
    this.email = By.name("email");
    this.confirmEmail = By.name("confirmemail");
    this.address = By.name("address");
    this.countryBox = By.className("select2-chosen");
    this.country = By.css("input.select2-input");
    this.additionalNotes = By.name("additionalnotes");
    this.confirmBookingLocator = By.name("guest");
};

BookingDetailsPage.prototype.setFirstName = function(fName) {
    this.driver.findElement(this.firstName).sendKeys(fName);
};

BookingDetailsPage.prototype.setLastName = function(lName) {
    this.driver.findElement(this.lastName).sendKeys(lName);
};

BookingDetailsPage.prototype.setPhone = function(phoneNumber) {
    this.driver.findElement(this.phone).sendKeys(phoneNumber);
};

BookingDetailsPage.prototype.setEmail = function(emailAddress) {
    this.driver.findElement(this.email).sendKeys(emailAddress);
};

BookingDetailsPage.prototype.setConfirmEmail = function(emailAddress) {
    this.driver.findElement(this.confirmEmail).sendKeys(emailAddress);
};

BookingDetailsPage.prototype.setAddress = function(address) {
    this.driver.findElement(this.address).sendKeys(address);
};

BookingDetailsPage.prototype.setAdditionalNotes = function(notes) {
    this.driver.findElement(this.additionalNotes).sendKeys(notes);
};

BookingDetailsPage.prototype.setCountry = function(country) {
    this.driver.findElement(this.countryBox).click();
    this.driver.findElement(this.country).sendKeys(country);
    this.driver.findElement(this.country).sendKeys(webDriver.Key.ENTER);
};


BookingDetailsPage.prototype.confirmBooking = function() {
    var theDriver = this.driver;
    var confirmBookingElement = this.driver.findElement(this.confirmBookingLocator);
    confirmBookingElement.isDisplayed().then(function(isDisplayed) {
        if (!isDisplayed) {
            var actions = new webDriver.ActionSequence(theDriver);
            actions.mouseMove(confirmBookingElement);
            actions.perform();
        }
    });
    confirmBookingElement.click();
    return new BookingConfirmationPage(this.driver);
};

module.exports = BookingDetailsPage;